Watch 401 response error globally and redirect to the reconnection page

eason9487 opened the pull request:
### Changes proposed in this Pull Request:

Implemented partial #486 and based on #833.

Please refer to the reconnection step 1 with the **Google account reconnection flow** section in the #486 description.

- Implement a middleware to handle error response instance globally
- Redirect to the reconnection page if receives any 401 response

### Screenshots:

![Kapture 2021-06-28 at 19 25 47](https://user-images.githubusercontent.com/17420811/123629749-3c3f6f80-d847-11eb-8b13-7b0bbf52dede.gif)


### Detailed test instructions:

1. Disconnect Google account from the Connection Test page
2. Go to the dashboard page or any other page which will request data through Google credentials
3. When receiving any 401 response error, the browser should be redirected to the reconnection page `/wp-admin/admin.php?page=wc-admin&subpath=%2Freconnect-accounts&path=%2Fgoogle%2Fsettings`

### Changelog entry

eason9487 committed:
Implement a middleware to handle response instance globally

eason9487 committed:
Redirect to reconnection page if receives any 401 response

eason9487 committed:
Add test cases for the createErrorResponseCatcher middleware

findingsimple commented:
Nice @eason9487

tomalec commented on js/src/data/api-fetch-middlewares.js:
```diff
@@ -0,0 +1,56 @@
+						return null;
+					}
+
+					return response.json
+						? response.json()
+						: Promise.reject( response );
```
:grey_question: 
This code looks very alike https://github.com/woocommerce/google-listings-and-ads/pull/834/files#diff-e8891a9027522176e1b7cb1d18e4e2373146aebf6d12733fdce179ae08337661R34-R36

If I understood correctly, here we are 
"Return parsed JSON if available, reject with response otherwise"
While L-34-36 does
"Parse the JSON, and reject with the result"

eason9487 replied:
> If I understood correctly, here we are
> "Return parsed JSON if available, reject with response otherwise"
> While L-34-36 does
> "Parse the JSON, and reject with the result"

Yes, these two code fragments work just as you understood.

tomalec commented on js/src/data/api-fetch-middlewares.test.js:
```diff
@@ -0,0 +1,105 @@
+
+	const passReject = ( response ) => Promise.reject( response );
+
+	const optionsDefaultParse = { path: `${ API_NAMESPACE }/hi` };
+	const optionsDontParse = { path: `${ API_NAMESPACE }/hi`, parse: false };
```
:nail_care: 
Maybe I'm paranoid, but I'd move those into `beforeEach` as we would re-use it in every test

eason9487 replied:
Updated by https://github.com/woocommerce/google-listings-and-ads/pull/834/commits/12e78f02d06534727fbcd6945590aa76d40f14bc

tomalec commented on js/src/data/api-fetch-middlewares.test.js:
```diff
@@ -0,0 +1,105 @@
+		it( 'should only handle requests where the path belongs to this plugin', async () => {
+			const data = {};
+			const optionsOtherNamespace = { path: '' };
+			const callback = jest.fn().mockImplementation( ( options ) => {
+				expect( options ).toBe( optionsOtherNamespace );
+				return data;
+			} );
```
:nail_care: Personally, I like to [name the mocks](https://jestjs.io/docs/mock-functions#mock-names), to make assertion errors more descriptive

eason9487 replied:
Updated by https://github.com/woocommerce/google-listings-and-ads/pull/834/commits/4ae6932816b8af405bea9e23869b1d9495c8b3c0

tomalec replied:
Sorry, but that's not what I meant.

By "name the mocks" I mean adding the name property to the mock object, not changing the variable name.

So instead of "anonymous" `mockCallback`
![anonymous mock](https://user-images.githubusercontent.com/17435/124480725-dce1e200-dda7-11eb-86ea-f450673f0632.png)

with `callback = jest.fn()/*...*/.mockName( "middleware's callback" );`
we can have, named `callback`
![image](https://user-images.githubusercontent.com/17435/124481070-31855d00-dda8-11eb-8559-da230a5de5f3.png)

(We can call it `.mockName( 'callback' )` as well, to match the variable name. "middleware" could be read from the test description.)

eason9487 replied:
Sorry, I didn't read the link's content clearly. Thanks for the elaborate explanation. After I took the https://github.com/woocommerce/google-listings-and-ads/pull/834/commits/f43e6f24d08e5861d9083f44a3dd08db3e1c657a you proposed, the `mockName` was chained to all the mock function calls. 👍

tomalec commented on js/src/data/api-fetch-middlewares.test.js:
```diff
@@ -0,0 +1,105 @@
+			middleware = createErrorResponseCatcher( passReject );
+		} );
+
+		it( 'should only handle requests where the path belongs to this plugin', async () => {
```
:question:
"Handle" is quite ambiguous. After looking at the implementation I'd expect that for other namespaces the callback should not be called at all. But here we are expecting it to be called

eason9487 replied:
I found the test case itself probably too ambiguous to know the test intention, so I made a change https://github.com/woocommerce/google-listings-and-ads/pull/834/commits/fbdc193f9b4c0370bd5b98765df6e5020dce92a0 to see if it more clear

tomalec commented on js/src/data/api-fetch-middlewares.test.js:
```diff
@@ -0,0 +1,105 @@
+				return data;
+			} );
+
+			const result = await middleware( optionsOtherNamespace, callback );
+			expect( callback ).toHaveBeenCalledTimes( 1 );
```
:nail_care: ^ :nail_care: - like totally opinionated, codestyle costmetic.

Personally, I like to make a vertical space between action and assertion.

eason9487 replied:
Adjusted by https://github.com/woocommerce/google-listings-and-ads/pull/834/commits/1ef2778926b9298c0a3b4deee3f4ac7fae9ad1d6

tomalec commented on js/src/data/api-fetch-middlewares.test.js:
```diff
@@ -0,0 +1,105 @@
+			const handler = createFetchHandler( 200, { data: 'hi' } );
+			const result = await middleware( optionsDontParse, handler );
+			expect( result.status ).toEqual( 200 );
+			expect( result.json ).toBeInstanceOf( Function );
```
:nail_care: 
"in a successful response" would suggest asserting `expect( result ).toBeInstanceOf( Response );` but I guess as it is mocked we don't have a constructor here.

eason9487 replied:
This test case intends to assert that the response instance should resolve directly without any parse process in this middleware. I made a change https://github.com/woocommerce/google-listings-and-ads/pull/834/commits/752411d4c5582485dd950f3526c1a609f63358a4 to check it directly.

And the real response instance is created and passed in from the outside of this middleware's scope. But I think we can rely on the implementation of `@wordpress/api-fetch` and believe it will pass the real response instance when the `parse` option is `false`.

eason9487 committed:
Move reused variables to beforeEach.

Address https://github.com/woocommerce/google-listings-and-ads/pull/834#discussion_r663177789

eason9487 committed:
Prefix "mock" to mock func to make the assertion errors more descriptive

Address https://github.com/woocommerce/google-listings-and-ads/pull/834#discussion_r663179303

eason9487 committed:
Make a vertical space between action and assertion

Address https://github.com/woocommerce/google-listings-and-ads/pull/834#discussion_r663228459

eason9487 committed:
Adjust the namespace matching test and make its description more clear

Address https://github.com/woocommerce/google-listings-and-ads/pull/834#discussion_r663227720

eason9487 committed:
Adjust the `parse: false` test cases to assert the response instance reference directly

Address https://github.com/woocommerce/google-listings-and-ads/pull/834#discussion_r663229777

tomalec commented on js/src/data/api-fetch-middlewares.js:
```diff
@@ -0,0 +1,56 @@
+
+		const { parse: shouldParseResponse = true } = options;
+
+		return next( { ...options, parse: false } )
+			.catch( onErrorResponse )
+			.catch( ( response ) => {
+				if ( shouldParseResponse && response.json ) {
+					return response
+						.json()
+						.then( ( errMessage ) => Promise.reject( errMessage ) );
+				}
+
+				throw response;
+			} )
+			.then( ( response ) => {
```
:nail_care: 
I think I found something that made me spend so much time parsing this in my head, trying to understand why this code is here and what it actually does. 

I hope the following comments could save that head-scratching for some other folks in the future.
```suggestion
		// Call onErrorResponse callback, then parse its fullfilment value to JSON if needed (and possible).
		return next( { ...options, parse: false } )
			.catch( onErrorResponse )
			.catch( ( response ) => {
				// Try parsing the rejection reason.
				// Reject with the response - parsed to JSON if required and possible.
				if ( shouldParseResponse && response.json ) {
					return response
						.json()
						.then( ( errMessage ) => Promise.reject( errMessage ) );
				}

				throw response;
			} )
			.then( ( response ) => {
				// Parse the "recovered" value.
```

tomalec replied:
Also, on my way to the discovery of what those functions do I refactored the code a little. To something I personally find more readable [`fix/486-watch-401-error-parsing-helper`](https://github.com/woocommerce/google-listings-and-ads/compare/fix/486-watch-401-error...fix/486-watch-401-error-parsing-helper?expand=1) feel free to ignore it or merge it if you find any value in it.

eason9487 replied:
Thanks, I merged https://github.com/woocommerce/google-listings-and-ads/pull/834/commits/ec2593e55cd7a192721574889e3cea2d88442dad and appended two commits, https://github.com/woocommerce/google-listings-and-ads/pull/834/commits/0017500856474f5fc1e766034bcdc7899e214bb2 https://github.com/woocommerce/google-listings-and-ads/pull/834/commits/d114b76c180ac6bdf68a84c451018cad32e4b7ad, to tweak code comments.

tomalec committed:
Move conditional JSON parsing to a helper function,

to make `createErrorResponseCatcher` code DRYier, and its logic more empahsized.
Add code comments to give more background on what's happening.
Related to https://github.com/woocommerce/google-listings-and-ads/pull/834/files#r664127329

tomalec committed:
Rephrase `createErrorResponseCatcher` tests' descriptions.

To hopefully make them more readable.
Add named mocks for more descriptive assertion errors. Addresses: https://github.com/woocommerce/google-listings-and-ads/pull/834#discussion_r663179303.

tomalec commented:
I had some comments for tests. Mostly related to descriptions. I'm not a plain-English expert, and those are quite opinionated personal preferences.

1. :question:  Looking at implementation I find ["force overwrite `parse` option to `false`"](https://github.com/woocommerce/google-listings-and-ads/pull/834/files#diff-76bfca5f4706db9dbf5fc54ba09c014f6a8a55196566e3fbf3e84c16f9a4b14eR61) not accurate and TBH do not understand "and return the same result". From what I see at the implementation. We [overwrite the default value with `true`](https://github.com/woocommerce/google-listings-and-ads/pull/834/files#diff-e8891a9027522176e1b7cb1d18e4e2373146aebf6d12733fdce179ae08337661R28), then we [force it to be `false`for the next middleware](https://github.com/woocommerce/google-listings-and-ads/pull/834/files#diff-e8891a9027522176e1b7cb1d18e4e2373146aebf6d12733fdce179ae08337661R30), then still continue to [respect the provided value in final fulfillment](https://github.com/woocommerce/google-listings-and-ads/pull/834/files#diff-e8891a9027522176e1b7cb1d18e4e2373146aebf6d12733fdce179ae08337661R33).

	Then even though the test description states that it's about forcing `false` (for next middleware) it actually tests default `true` for a callback, without asserting anything on middleware. Which looks to me as a :construction: testing something else than described.

2. :nail_care:  Personally, I prefer the test descriptions where the conditions come first, and the expectation is last. I find it easier to write especially in nested `describe`s and more importantly easier to read when the full desc is build:

```js
describe( `When the path…`, () => {
	describe( 'for successful…', () => {
		it( 'should call …', () => {
```

3. :nail_care: Plus we can use capital letters to build sentences that indicate the granularity/detail level of our test case.

This lets us build descriptions like:

> createErrorResponseCatcher > When the path belongs to this plugin's namespace, should overwrite default value of `parse` option to `false` and return the same result > When given `parse: false` option, for a successful response, should resolve with the response instance.

Here is the example of proposed alternatives, feel free to merge it or ignore it :)
[`fix/486-watch-401-error-test-desc](https://github.com/woocommerce/google-listings-and-ads/compare/fix/486-watch-401-error...fix/486-watch-401-error-test-desc?expand=1)

tomalec reviewed (commented):
Tested locally and reviewed the code. Left few cosmetic comments. Functionalwise LGTM, but I found one thing that makes me think tests describes something else than they actually test (see [above](https://github.com/woocommerce/google-listings-and-ads/pull/834#issuecomment-874350184))

tomalec committed:
Fix tests descriptions, test forcing `parse` for a next middleware.

Addresses https://github.com/woocommerce/google-listings-and-ads/pull/834#issuecomment-874350184.

eason9487 committed:
Remove async from responseOrJSON to fit with its JSDoc

eason9487 committed:
Tweak some comments in the api-fetch-middlewares.js

eason9487 committed:
Merge branch 'fix/486-watch-401-error-test-desc' into fix/486-watch-401-error

eason9487 committed:
Add a comment for why the default parse option is true

eason9487 commented:
> I found one thing that makes me think tests describes something else than they actually test (see [above](https://github.com/woocommerce/google-listings-and-ads/pull/834#issuecomment-874350184))

@tomalec, thank you for your thorough reviews! I think because I read the `api-fetch` implementation before developing this middleware, and I skipped these context during development, causing blind spots. And it leads others to lose the context when reading code. I merged the **fix/486-watch-401-error-test-desc** branch, and the test description and structure are getting more clear. 👍 

This PR is ready for the next round of review 🙏

tomalec commented on js/src/data/index.js:
```diff
@@ -21,6 +25,17 @@ registerStore( STORE_KEY, {
+apiFetch.use(
+	createErrorResponseCatcher( ( response ) => {
+		if ( response.status === 401 ) {
+			getHistory().replace( getReconnectAccountsUrl() );
```
:scroll:
I have a doubt about the UX here. Do we really want to replace the history = disallow a user to go back to retry after they re-connect the account?

Personally, as a heavy back button user, I'd prefer `pushState` here

eason9487 replied:
Currently, when going back to the page which had got error responses, we don't have a wide implementation to retry loading data. Thus, there's no API request to trigger 401 error response and re-redirect user to reconnection page again. It would display a page without any error message (#849) and no available data to render. Compare to that, I think it should be relatively acceptable to use `replace` for now?

tomalec replied:
I didn't mean retry loading data. The flow I was referring to was:
1. User: "I'd like to go to the dashboard to see my stats"
2. User: visits `/wp-admin/admin.php?page=wc-admin&path=%2Fgoogle%2Fdashboard`
2. GLA: Fetch the stats, 401, redirects to reconnect page
4. ⋮
5. User: reconnects Google  account
6. GLA: "Account successfully reconnected"
7. User: "Cool, I'd like to get back to where I was at" Clicks "back button" n times/right-click to pick from history.

> Compare to that, I think it should be relatively acceptable to use `replace` for now?

Definitely, probably the flow could change in the meantime anyway. And `replace` is better than the empty page.

eason9487 replied:
> 1\. User: "I'd like to go to the dashboard to see my stats"
> 6\. GLA: "Account successfully reconnected"

After merging #838, users will be redirected to the dashboard page at step 6. But no matter the original page users wanted to view at step 1, they're always redirected to the dashboard page after reconnected.

I was thinking about adding a UX improvement by:

1. When the 401 is detected, store the current page path in `sessionStorage` before redirecting to reconnect
2. After Google account is successfully reconnected, bring users back to the page stored in `sessionStorage` if it exists. Otherwise, bring them to the default (dashboard) page.

By this flow, it'd still using `replace` but the concern you mentioned could be handled in a more user-friendly way. WDYT @tomalec ?

tomalec reviewed (approved):
LGTM thanks! :)
Left one note about the UX https://github.com/woocommerce/google-listings-and-ads/pull/834#discussion_r664373065
